### ISSUE TITLE
How to Do Code Reviews Like a Human (Part One)

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7389]
How to Do Code Reviews Like a Human (Part One)<br>
<br>
https://ift.tt/2yfY1yv<br>
<br>
Michael Lynch<br>
<br>
<div>
<div>
<div>
<div>
<p>Lately, I&rsquo;ve been reading articles about best practices for code reviews. I notice that these articles focus on finding bugs to the exclusion of almost every other component of a review. Communicating issues you discover in a constructive and professional way? Irrelevant! Just identify all the bugs, and the rest will take care of itself.</p>
<p>So I had a revelation: if this works for code, why not romance? With that, I&rsquo;m announcing my new ebook to help developers with their love lives:</p>
<p>My revolutionary ebook teaches you <strong>proven techniques</strong> for maximizing the number of deficiencies you find in your partner. The ebook does <strong>not</strong> cover:</p>
<ul>
<li>Communicating issues to your partner with empathy and understanding.</li>
<li>Helping your partner address their weaknesses.</li>
</ul>
<p>Based on my reading of code review literature, those parts of a relationship are <em>obvious</em> and <em>not worth discussing</em>.</p>
<p>Does this sound like a good ebook to you? I&rsquo;m assuming you just yipped &ldquo;Nonononono!&rdquo;</p>
<p>So, why is that the way we talk about code reviews?</p>
<p>I can only assume the articles I&rsquo;ve read are from the future, where all developers are robots. In that world, your teammates welcome thoughtlessly-worded critiques of their code because processing such information warms their cold, robot hearts.</p>
<p>I&rsquo;m going to make the bold assumption that you want to improve code reviews in the present, where your teammates are humans. I&rsquo;ll make the even bolder assumption that a positive relationship with your colleagues is an end in itself and not simply a variable you adjust to minimize your cost-per-defect. How would your review practices change under these circumstances?</p>
<p>In this article, I discuss techniques that treat the code review as not only a technical process but a social one as well.</p>
<h2>What is a code review? <a href="https://mtlynch.io/human-code-reviews-1/#what-is-a-code-review">&#128279;&#65038;</a>
</h2>
<p>The term &ldquo;code review&rdquo; can refer to a range of activities, from simply reading some code over your teammate&rsquo;s shoulder to a 20-person meeting where you dissect code line by line. I use the term to refer to a process that&rsquo;s formal and written, but not so heavyweight as a series of in-person code inspection meetings.</p>
<p>The participants in a code review are the <strong>author</strong>, who writes the code and sends it for review, and the <strong>reviewer</strong>, who reads the code and decides when it&rsquo;s ready to be merged in to the team&rsquo;s codebase. A review can have multiple reviewers, but I assume for simplicity that you are the sole reviewer.</p>
<p>Before the code review begins, the author must create a <strong>changelist</strong>. This is a set of changes to source code that the author wants to merge in to the team&rsquo;s codebase.</p>
<p>A review begins when the author sends their changelist to the reviewer. Code reviews happen in <strong>rounds</strong>. Each round is one complete round-trip between the author and reviewer: the author sends changes, and the reviewer responds with written feedback on those changes. Every code review has one or more rounds.</p>
<p>The review ends when the reviewer <strong>approves</strong> the changes. This is commonly referred to as giving LGTM, shorthand for &ldquo;looks good to me.&rdquo;</p>
<h2>Why is this hard? <a href="https://mtlynch.io/human-code-reviews-1/#why-is-this-hard">&#128279;&#65038;</a>
</h2>
<p>If a programmer sends you a changelist that they think is awesome, and you write them an extensive list of reasons why it&rsquo;s not, that&rsquo;s a sensitive message to get across.</p>
<blockquote>
<p>That&rsquo;s one reason I don&rsquo;t miss IT, because programmers are very unlikable people&hellip; In aviation, for example, people who greatly overestimate their level of skill are all dead.</p>
<p>-Philip Greenspun, co-founder of ArsDigita, excerpted from <a href="http://amzn.to/2g6oTsV"><em>Founders at Work</em></a></p>
</blockquote>
<p>It&rsquo;s easy for an author to interpret criticism of their code as an implication that they are an incompetent programmer. Code reviews are an opportunity to share knowledge and make informed engineering decisions. But that can&rsquo;t happen if the author perceives the discussion as a personal attack.</p>
<p>As if this wasn&rsquo;t difficult enough, you also have the challenge of conveying your thoughts in writing, where the risk of miscommunication is higher. The author can&rsquo;t hear your tone of voice or see your body language, so it&rsquo;s even more important to articulate your feedback carefully. To an author who&rsquo;s feeling defensive, an innocuous note like, &ldquo;You forgot to close the file handle,&rdquo; can read as, &ldquo;I can&rsquo;t <em>believe</em> you forgot to close the file handle! You&rsquo;re such an idiot.&rdquo;</p>
<h2>Techniques <a href="https://mtlynch.io/human-code-reviews-1/#techniques">&#128279;&#65038;</a>
</h2>
<ol>
<li><a href="https://mtlynch.io/human-code-reviews-1/#let-computers-do-the-boring-parts">Let computers do the boring parts</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#settle-style-arguments-with-a-style-guide">Settle style arguments with a style guide</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#start-reviewing-immediately">Start reviewing immediately</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#start-high-level-and-work-your-way-down">Start high level and work your way down</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#be-generous-with-code-examples">Be generous with code examples</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#never-say-you">Never say &ldquo;you&rdquo;</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#frame-feedback-as-requests-not-commands">Frame feedback as requests, not commands</a></li>
<li><a href="https://mtlynch.io/human-code-reviews-1/#tie-notes-to-principles-not-opinions">Tie notes to principles, not opinions</a></li>
</ol>
<h3>Let computers do the boring parts <a href="https://mtlynch.io/human-code-reviews-1/#let-computers-do-the-boring-parts">&#128279;&#65038;</a>
</h3>
<p>Between interruptions like meetings and emails, the time you have available to focus on code is scarce. Your mental stamina is in even shorter supply. Reading a teammate&rsquo;s code is cognitively taxing and requires a high level of concentration. Don&rsquo;t squander these resources on tasks a computer can do, especially when a computer can do them better.</p>
<p>Whitespace errors are an obvious example. Compare how much effort it takes for a human reviewer to find an indenting mistake and work with the author to correct it as opposed to just using an automated formatting tool:</p>
<table>
<thead>
<tr>
<th>Effort required with a human reviewer</th>
<th>Effort required with a formatting tool</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<ol>
<li>Reviewer searches for whitespace issues and finds incorrect indentation.</li>
<li>Reviewer writes a note calling out the incorrect indentation.</li>
<li>Reviewer rereads their note to make sure that it's worded in a clear, non-accusatory way.</li>
<li>Author reads the note.</li>
<li>Author corrects the code indentation.</li>
<li>Reviewer verifies that the author addressed their note properly.</li>
</ol>
</td>
<td>Nothing!</td>
</tr>
</tbody>
</table>
<p>The right side is empty because the author uses a code editor that automatically formats the whitespace every time they hit &ldquo;Save.&rdquo; At worst, the author sends their code out for review, and the <a href="https://en.wikipedia.org/wiki/Continuous_integration">continuous integration</a> solution reports that the whitespace is incorrect. The author fixes the issue without the reviewer ever having to care.</p>
<p>Look for mechanical tasks in your code reviews that you can automate away. Here are the common ones:</p>
<table>
<thead>
<tr>
<th>Task</th>
<th>Automated solution</th>
</tr>
</thead>
<tbody>
<tr>
<td>Verify the code builds</td>
<td>Continuous integration solution, such as <a href="https://travis-ci.com">Travis</a> or <a href="https://circleci.com/">CircleCI</a>.</td>
</tr>
<tr>
<td>Verify automated tests pass</td>
<td>Continuous integration solution, such as <a href="https://travis-ci.com">Travis</a> or <a href="https://circleci.com/">CircleCI</a>.</td>
</tr>
<tr>
<td>Verify code whitespace matches team style</td>
<td>Code formatter, such as <a href="https://clang.llvm.org/docs/ClangFormat.html">ClangFormat</a> (C/C++ formatter) or <a href="https://golang.org/cmd/gofmt/">gofmt</a> (Go formatter).</td>
</tr>
<tr>
<td>Identify unused imports or unused variables</td>
<td>Code linters, such as <a href="https://pypi.python.org/pypi/pyflakes">pyflakes</a> (Python linter) or <a href="http://jslint.com/help.html">JSLint</a> (JavaScript linter).</td>
</tr>
</tbody>
</table>
<p>Automation helps you make more meaningful contributions as a reviewer. When you can ignore a whole class of issues, such as the ordering of <code>imports</code> or naming conventions for source filenames, it allows you to focus on more interesting things like functional errors or weaknesses in readability.</p>
<p>Automation benefits the author as well. It allows them to discover careless mistakes in seconds instead of hours. The instant feedback makes it easier to learn from and cheaper to fix because the author still has the relevant context in their head. Plus, if they have to hear about a dumb mistake they made, it&rsquo;s much easier on their ego if they hear it from a computer instead of from you.</p>
<p>Work with your team to build these automated checks directly into the code review workflow (e.g., <a href="https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks">pre-commit hooks</a> in Git or <a href="https://developer.github.com/webhooks/">webhooks</a> in Github). If the review process requires the author to run these checks manually, you forfeit most of the benefit. The author will invariably forget on occasion which forces you to continue reviewing for the simple issues that automation is meant to handle instead.</p>
<h3>Settle style arguments with a style guide <a href="https://mtlynch.io/human-code-reviews-1/#settle-style-arguments-with-a-style-guide">&#128279;&#65038;</a>
</h3>
<p>Arguments about style are a waste of time in reviews. Consistent style is certainly important, but a code review is not the time to bicker about where to put the curly braces. The best way to excise style debates from your reviews is by keeping a style guide.</p>
<p>A good style guide defines not only superficial elements like naming conventions or whitespace rules but also how to use the features of the given programming language. JavaScript and Perl, for example, are packed with functionality &mdash; they offer many ways to implement the same logic. A style guide defines The One True Way of doing things so that you don&rsquo;t end up with half your team using one set of language features while the other half uses a totally different set of features.</p>
<p>Once you have a style guide, you don&rsquo;t have to waste review cycles arguing with the author about whose naming conventions are best. Just defer to the style guide and move on. If your style guide doesn&rsquo;t specify a convention about a particular issue, it&rsquo;s generally not worth arguing about. If you encounter a style issue your guide doesn&rsquo;t cover and it&rsquo;s important enough to discuss, hash it out with your team. Then, record the decision in your style guide so you never have to have that discussion again.</p>
<p><em><strong>Option 1: Adopt an existing style guide</strong></em></p>
<p>If you search online, you can find published style guides ripe for the taking. <a href="https://google.github.io/styleguide/">Google&rsquo;s style guides</a> are the most well-known, but you can find others if this style doesn&rsquo;t suit you. By adopting an existing guide, you inherit the benefits of a style guide without the substantial costs of creating one from scratch.</p>
<p>The downside is that organizations optimize their style guides for their own particular needs. For example, Google&rsquo;s style guides are conservative about <a href="https://google.github.io/styleguide/cppguide.html#C++11">using new language features</a> because they have an enormous codebase with code that has to run on everything from a home router to the latest iPhone. If you&rsquo;re a four-person startup with a single product, you may choose to be more aggressive in using cutting-edge language features or extensions.</p>
<p><em><strong>Option 2: Create your own style guide incrementally</strong></em></p>
<p>If you don&rsquo;t want to adopt an existing guide, you can create your own. Every time a style argument arises during a code review, raise the question to your whole team to decide what the official convention should be. When you reach agreement, codify that decision in your style guide.</p>
<p>I prefer to keep my team&rsquo;s style guide as Markdown under source control (e.g., <a href="https://pages.github.com/">GitHub pages</a>). That way, any changes to the style guide go through the normal review process &mdash; someone has to explicitly approve the change, and everyone on the team has a chance to raise concerns. Wikis and Google Docs are acceptable options as well.</p>
<p><em><strong>Option 3: The hybrid approach</strong></em></p>
<p>By combining options 1 and 2, you can adopt an existing style guide as your base, and then maintain a local style guide to extend or override the base. A good example of this is the <a href="https://chromium.googlesource.com/chromium/src/+/master/styleguide/c++/c++.md">Chromium C++ style guide</a>. It uses <a href="https://google.github.io/styleguide/cppguide.html">Google&rsquo;s C++ style guide</a> as a base, but makes its own changes and additions on top of it.</p>
<p>Treat code reviews as a high priority. When you&rsquo;re actually reading the code and giving feedback, take your time, but <em>start</em> your review immediately &mdash; ideally, within minutes.</p>
<p>If a teammate sends you a changelist, it likely means that they are blocked on other work until your review is complete. In theory, source control systems allow the author to branch, continue working, and then forward-merge changes from the review into their new branch. In reality, there are about four developers total who can do that efficiently. It takes everyone else so long to untangle three-way diffs that it can cancel out any progress made waiting for the review to come back.</p>
<p>When you start reviews immediately, you create a virtuous cycle. Your review turnaround becomes purely a function of the size and complexity of the author&rsquo;s changelist. This incentivizes authors to send small, narrowly-scoped changelists. These are easier and more pleasant for you to review, so you review them faster, and the cycle continues.</p>
<p>Imagine that your teammate implements a new feature that requires 1,000 lines of code changes. If they know you can review a 200-line changelist in about 2 hours, they can break their feature into changelists of about 200 lines each and get the whole feature checked in within a day or two. If, however, you take a day to do all code reviews, regardless of size, now it takes a week to get that feature checked in. Your teammate doesn&rsquo;t want to sit around for a week, so they&rsquo;re incentivized to send larger code reviews, like 500-600 lines each. These are more costly to review and yield poorer feedback because it&rsquo;s more difficult to keep context on a 600-line change than a 200-line change.</p>
<p>The absolute maximum turnaround on a review round should be one business day. If you&rsquo;re struggling with a higher-priority issue and can&rsquo;t complete a round of review in under a day, let your teammate know and give them the opportunity to reassign it to someone else. If you&rsquo;re forced to decline reviews more than about once per month, it likely means that your team needs to reduce its pace so that you can maintain sane development practices.</p>
<h3>Start high level and work your way down <a href="https://mtlynch.io/human-code-reviews-1/#start-high-level-and-work-your-way-down">&#128279;&#65038;</a>
</h3>
<p>The more notes you write in a given review round, the more you risk making the author feel overwhelmed. The exact limit varies by developer, but the danger zone generally begins in the range of 20-50 notes in a single round of review.</p>
<p>If you&rsquo;re worried about drowning the author in a sea of notes, restrict yourself to high-level feedback in the early rounds. Focus on issues like redesigning a class interface or splitting up complex functions. Wait until those issues are resolved before tackling lower-level issues, such as variable naming or clarity of code comments.</p>
<p>Your low-level notes might become moot once the author integrates your high-level notes. By deferring them to a later round, you save yourself the nontrivial work of writing carefully-worded comments calling out the issues, and you spare the author from processing unnecessary notes. This technique also segments the layers of abstraction you focus on during the review, helping you and the author work through the changelist in a clear, systematic way.</p>
<h3>Be generous with code examples <a href="https://mtlynch.io/human-code-reviews-1/#be-generous-with-code-examples">&#128279;&#65038;</a>
</h3>
<p>In an ideal world, the code author would be thankful for every review they receive. It&rsquo;s an opportunity for them to learn, and it protects them from mistakes. In reality, there are a number of external factors that could cause the author to perceive the review negatively and resent you for giving them notes. Maybe they&rsquo;re under pressure to meet a deadline, so anything other than your instant, rubber-stamp approval feels like obstruction. Maybe you haven&rsquo;t worked together much, so they don&rsquo;t trust that your feedback is well-intentioned.</p>
<p>A great way to make an author feel good about the review process is to find opportunities to give them gifts during the review. And what&rsquo;s the gift all developers love to receive? Code examples, of course.</p>
<p>If you lighten the author&rsquo;s load by writing out some of the changes you&rsquo;re suggesting, you demonstrate that you are generous with your time as a reviewer.</p>
<p>For example, imagine that you have a colleague who is not familiar with the <a href="http://treyhunner.com/2015/12/python-list-comprehensions-now-in-color/">list comprehensions</a> feature of Python. They send you a code review that includes these lines:</p>
<div>
<pre><code>urls = []
<span>for</span> path <span>in</span> paths:
  url = <span>'https://'</span>
  url += domain
  url += path
  urls.append(url)
</code></pre>
</div>
<p>Responding, &ldquo;Can we simplify this with a list comprehension?&rdquo; will annoy them because now they have to spend 20 minutes researching something they&rsquo;ve never used before.</p>
<p>They will be much happier to receive a note like the following:</p>
<blockquote>
<p>Consider simplifying with a list comprehension like this:</p>
<div>
<pre><code>urls = [<span>'https://'</span> + domain + path <span>for</span> path <span>in</span> paths]
</code></pre>
</div>
</blockquote>
<p>This technique is not limited to one-liners. I&rsquo;ll often create my own branch of the code to demonstrate a large proof of concept to the author, such as breaking up a large function or adding a unit test to cover an additional edge case.</p>
<p>Reserve this technique for clear, uncontroversial improvements. In the list comprehension example above, few developers would object to an 83% reduction in lines of code. In contrast, if you write a lengthy example to demonstrate a change that is &ldquo;better&rdquo; based on your own personal taste (e.g., style changes), code examples make you look pushy instead of generous.</p>
<p>Limit yourself to two or three code examples per review round. If you start writing the author&rsquo;s whole changelist for them, it signals that you don&rsquo;t think they&rsquo;re capable of writing their own code.</p>
<h3>Never say &ldquo;you&rdquo; <a href="https://mtlynch.io/human-code-reviews-1/#never-say-you">&#128279;&#65038;</a>
</h3>
<p>This one is going to sound weird, but hear me out: never use the word &ldquo;you&rdquo; in a code review.</p>
<p>The decisions you reach in a review should be based on what makes the code better rather than who came up with the idea. Your teammate put significant effort into their changelist and is likely proud of the work they did. Their natural reaction to hearing criticism of their work is to feel defensive and protective.</p>
<p>Word your feedback in a way that minimizes the risk of raising your teammate&rsquo;s defenses. Be clear that you&rsquo;re critiquing the code, not the coder. When an author sees &ldquo;you&rdquo; in a comment, it brings their focus away from the code and back to themselves. This increases the risk that they&rsquo;ll take your criticism personally.</p>
<p>Consider this harmless comment:</p>
<blockquote>
<p>You misspelled &lsquo;successfully.&rsquo;</p>
</blockquote>
<p>The author can interpret that note in two very different ways:</p>
<ul>
<li>
<strong>Interpretation 1</strong>: Hey, good buddy! You misspelled &lsquo;successfully.&rsquo; But I still think you&rsquo;re smart! It was probably just a typo.</li>
<li>
<strong>Interpretation 2</strong>: You misspelled &lsquo;successfully,&rsquo; dumbass.</li>
</ul>
<p>Contrast this with a note that omits &ldquo;you&rdquo;:</p>
<blockquote>
<p>sucessfully -&gt; successfully</p>
</blockquote>
<p>The latter note is a simple correction and not a judgment of the author.</p>
<p>Fortunately, it&rsquo;s easy to rewrite your feedback to avoid the word &ldquo;you.&rdquo;</p>
<p><em><strong>Option 1: Replace &lsquo;you&rsquo; with &lsquo;we&rsquo;</strong></em></p>
<blockquote>
<p>Can <strong>you</strong> rename this variable to something more descriptive, like <code>seconds_remaining</code>?</p>
</blockquote>
<p>becomes:</p>
<blockquote>
<p>Can <strong>we</strong> rename this variable to something more descriptive, like <code>seconds_remaining</code>?</p>
</blockquote>
<p>&ldquo;We&rdquo; reinforces the team&rsquo;s collective responsibility for the code. The author may move on to a different company and so might you, but the team who owns this code will remain in one form or another. It can sound silly to say &ldquo;we&rdquo; when it&rsquo;s clearly something you expect the author to do themselves, but silly is better than accusatory.</p>
<p><em><strong>Option 2: Remove the subject from the sentence</strong></em></p>
<p>Another way to avoid using &ldquo;you&rdquo; is to use a shorthand that omits the subject from the sentence:</p>
<blockquote>
<p><strong>Suggest renaming</strong> to something more descriptive, like <code>seconds_remaining</code>.</p>
</blockquote>
<p>You can achieve a similar effect with the <a href="https://en.wikipedia.org/wiki/English_passive_voice">passive voice</a>. I generally avoid the passive voice like the plague in my technical writing, but it can be a helpful way of writing around &ldquo;you&rdquo;:</p>
<blockquote>
<p>This variable <strong>should be renamed</strong> to something more descriptive, like <code>seconds_remaining</code>.</p>
</blockquote>
<p>An additional option is to phrase it as a question, beginning with &ldquo;what about&hellip;&rdquo; or &ldquo;how about&hellip;":</p>
<blockquote>
<p><strong>What about renaming</strong> this variable to something more descriptive, like <code>seconds_remaining</code>?</p>
</blockquote>
<h3>Frame feedback as requests, not commands <a href="https://mtlynch.io/human-code-reviews-1/#frame-feedback-as-requests-not-commands">&#128279;&#65038;</a>
</h3>
<p>Code reviews require more tact and care than usual communication because there&rsquo;s a high risk of derailing the discussion into a personal argument. You would expect reviewers to dial up their politeness in reviews, but bizarrely I&rsquo;ve found them to go the opposite direction. Most people never say to a co-worker, &ldquo;Hand me that stapler, then fetch me a soda.&rdquo; But I&rsquo;ve seen numerous reviewers frame feedback with similarly pushy commands, such as, &ldquo;Move this class to a separate file.&rdquo;</p>
<p>Err on the side of being annoyingly gentle in your feedback. Frame your notes as requests or suggestions, not commands.</p>
<p>Compare the same note framed in two different ways:</p>
<table>
<thead>
<tr>
<th>Feedback framed as command</th>
<th>Feedback framed as request</th>
</tr>
</thead>
<tbody>
<tr>
<td>Move the <code>Foo</code> class to a separate file.</td>
<td>Can we move the <code>Foo</code> class to a separate file?</td>
</tr>
</tbody>
</table>
<p>People like to feel in control of their own work. Making a request of the author gives them a sense of autonomy.</p>
<p>Requests also make it easier for the author to push back politely. Maybe they have a good reason for their choice. If you frame your feedback as a command, any pushback from the author comes across as disobedience. If you frame your feedback as a request or a question, the author can simply answer you.</p>
<p>Compare how combative the conversation seems depending on how the reviewer frames their initial note:</p>
<table>
<thead>
<tr>
<th>Feedback framed as command (Combative)</th>
<th>Feedback framed as request (Cooperative)</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<strong>Reviewer</strong>: Move the <code>Foo</code> class to a separate file.<br>
<strong>Author</strong>: I don&rsquo;t want to do that because then it&rsquo;s far away from the <code>Bar</code> class. Clients will almost always use the two together.</td>
<td>
<strong>Reviewer</strong>: Can we move the <code>Foo</code> class to a separate file?<br>
<strong>Author</strong>: We could, but then it&rsquo;s far away from the <code>Bar</code> class, and clients will generally use these two classes together. What do you think?</td>
</tr>
</tbody>
</table>
<p>See how much more civil the conversation becomes when you <del>construct imaginary dialog to prove your point</del> frame your notes as requests instead of commands?</p>
<h3>Tie notes to principles, not opinions <a href="https://mtlynch.io/human-code-reviews-1/#tie-notes-to-principles-not-opinions">&#128279;&#65038;</a>
</h3>
<p>When you give the author a note, explain both your suggested change and the <em>reason</em> for the change. Instead of saying, &ldquo;We should split this class into two,&rdquo; it&rsquo;s better to say, &ldquo;Right now, this class is responsible for both downloading the file and parsing it. We should split it up into a downloader class and parsing class per the <a href="https://en.wikipedia.org/wiki/Single_responsibility_principle">single responsibility principle</a>.&rdquo;</p>
<p>Grounding your notes in principles frames the discussion in a constructive way. When you cite a specific reason, like, &ldquo;We should make this function private to minimize the class&rsquo; public interface,&rdquo; the author can&rsquo;t simply respond, &ldquo;No, I prefer it my way.&rdquo; Or rather, they <em>can</em>, but it would look silly because you demonstrated how the change satisfies a goal, and they just stated a preference.</p>
<p>Software development is both an art and science. You can&rsquo;t always articulate exactly what is wrong with a piece of code in terms of established principles. Sometimes code is just ugly or unintuitive, and it&rsquo;s hard to pin down why. In these cases, explain what you can, but keep it objective. If you say, &ldquo;<strong>I</strong> found this hard to understand,&rdquo; that&rsquo;s at least an objective statement, as opposed to, &ldquo;<strong>this is</strong> confusing,&rdquo; which is a value judgment and may not be true for every person.</p>
<p>Provide supporting evidence where possible in the form of links. The relevant section of your team&rsquo;s style guide is the best link you can provide. You can also link to documentation for the language or library. Highly-upvoted <a href="https://stackoverflow.com">StackOverflow</a> answers can work as well, but the farther you stray from authoritative documentation, the shakier your evidence becomes.</p>
<h2>Part two <a href="https://mtlynch.io/human-code-reviews-1/#part-two">&#128279;&#65038;</a>
</h2>
<p>If you enjoyed this post, check out <a href="https://mtlynch.io/human-code-reviews-2/">the second half of this article</a>, which focuses on bringing reviews to a successful close without ugly conflict. It includes techniques for:</p>
<ul>
<li>Handling excessively large code reviews,</li>
<li>Recognizing opportunities to give praise,</li>
<li>Respecting the scope of a review, and</li>
<li>Mitigating stalemates.</li>
</ul>
<hr>
<p><em>Edited by <a href="https://www.samanthamasonfreelancer.com">Samantha Mason</a>. Illustrations by <a href="https://www.linkedin.com/in/lolo-ology/">Loraine Yow</a>. Thanks to <a href="https://twitter.com/global4g">@global4g</a> for providing valuable feedback on an early draft of this post.</em></p>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via mtlynch.io<br>
<br>
December 9, 2020 at 08:34PM
